### PR TITLE
Add support for Jetty 12

### PR DESCRIPTION
**Pull request by obenkenobi**
### Overview
Adds instrumentation for Jetty 12. 

Jetty 12 follows a philosophy of being *Servlet API* independent. As a result Jetty now ships with a Core Server that makes no reference to the Servlet API. In addition, Jetty now supports every Jakarta EE spec from 8 to 10. As a result, an instrumentation module is needed for each EE implementation. This PR also adds instrumentation for the core Jetty Server. As a result 5 new instrumentation modules are added.  These new modules can be described as follows:

#### jetty-12 
This module instruments the core Jetty server that handles tasks that does not require the Servlet API. Since async activity is not managed in the core server, transactions cannot be started and managed in this module. The tasks this module does include:

- Reporting thread-pool metrics
- Reporting the dispatcher name and version as well as the port number for the server.
- Adding outbound CAT headers for outgoing requests. CAT however is deprecated during the time this module is made (i.e. December 2023).

#### jetty-ee*-servlet-12
This refers to any module named based on the pattern `jetty-ee*-servlet-12`.
These modules instrument the Servlet API implementations for Jetty.
As a result, the instrumentation works with tasks that require us to work with the Servlet API such as:

- Start and ending web transactions
- Linking transactions to async activity
- Reporting errors

#### List of new instrumentation modules
- jetty-12 
- jetty-ee8-servlet-12
- jetty-ee9-servlet-12
- jetty-ee10-servlet-12
- jetty-ee10-servlet-12.0.1

### Related Github Issue

https://github.com/newrelic/newrelic-java-agent/issues/1300

### CODE REVIEW

File: instrumentation/jetty-12/src/main/java/com/nr/agent/instrumentation/jetty12/JettySampler.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+        final ThreadPool threadPool = server.getThreadPool();
+        if (threadPool != null) {
+            final String name = threadPool.getClass().getSimpleName();
+
+            int totalThreads, busyThreads;
+
+            int threadCount = threadPool.getThreads();
+            int idleCount = threadPool.getIdleThreads();
+
+            // Jetty9 ExecutorThreadPool doesn't expose getMaxThreads(), so just approximate usage
```
Comment (jtduffy) [inline instrumentation/jetty-12/src/main/java/com/nr/agent/instrumentation/jetty12/JettySampler.java:62]: Jetty9 --> Jetty12?

File: instrumentation/jetty-12/src/main/java/org/eclipse/jetty/server/Response_Instrumentation.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+@Weave(originalName = "org.eclipse.jetty.server.Response", type = MatchType.Interface)
+public class Response_Instrumentation {
+    public void write(boolean last, ByteBuffer content, Callback callback) {
+        if (!isCommitted()) {
+            NewRelic.getAgent().getTransaction().addOutboundResponseHeaders();
+        }
+        Weaver.callOriginal();
+    }
+
+    public boolean isCommitted() {
```
Comment (jtduffy) [inline instrumentation/jetty-12/src/main/java/org/eclipse/jetty/server/Response_Instrumentation.java:21]: To reduce some instrumentation overhead, you can make your class abstract, and then declare `isCommitted` as abstract.
`public abstract boolean isCommitted();`

File: instrumentation/jetty-12/src/main/java/org/eclipse/jetty/server/Server_Instrumentation.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+public abstract class Server_Instrumentation {
+
+    protected void doStart() {
+        AgentBridge.privateApi.addSampler(new JettySampler(this), 1, TimeUnit.MINUTES);
+        Weaver.callOriginal();
+    }
+
+    // Required so that earlier jetty versions do not apply.
+    // Transactions are managed in the jetty-12-ee* modules.
+    public boolean handle(Request request, Response response, Callback callback) {
```
Comment (jtduffy) [inline instrumentation/jetty-12/src/main/java/org/eclipse/jetty/server/Server_Instrumentation.java:23]: Same here.. You can make the following methods `abstract`

**Review by jtduffy — CHANGES_REQUESTED** [review]
Looks good. Aside from my inline comments:
- Most of the class files are missing the copyright header. Those that have the header have the incorrect year (2022).

File: instrumentation/jetty-12/src/main/java/com/nr/agent/instrumentation/jetty12/JettySampler.java (hunk 1 above)
Comment (obenkenobi) [reply instrumentation/jetty-12/src/main/java/com/nr/agent/instrumentation/jetty12/JettySampler.java]: Changed it to `Jetty 9+`

**Comment by obenkenobi** [comment]
> Looks good. Aside from my inline comments:
> 
> * Most of the class files are missing the copyright header. Those that have the header have the incorrect year (2022).

Added the copyright headers

**Review by jtduffy — APPROVED** [review]

**Pull request merged**